### PR TITLE
add seaweedfs

### PR DESCRIPTION
**Pull request by chrislusf**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -142,7 +142,8 @@ LICENSE file.
     <redis.version>2.9.0</redis.version>
     <riak.version>2.0.5</riak.version>
     <rocksdb.version>6.2.2</rocksdb.version>
-    <s3.version>1.10.20</s3.version>
+    <s3.version>1.11.820</s3.version>
```
Comment (busbey) [inline pom.xml:145]: put the s3 update in its own PR?

File: seaweedfs/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+<!--
+Copyright (c) 2015 YCSB contributors. All rights reserved.
```
Comment (busbey) [inline seaweedfs/README.md:2]: copyright date is wrong

File: seaweedfs/README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing
+permissions and limitations under the License. See accompanying
+LICENSE file.
+-->
+Quick Start
+===============
```
Comment (busbey) [inline seaweedfs/README.md:18]: A pointer to SeaweedFS intro docs?

File: seaweedfs/pom.xml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- 
+Copyright (c) 2015-2016 YCSB contributors. All rights reserved.
```
Comment (busbey) [inline seaweedfs/pom.xml:3]: copyright date is wrong

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ * <p>
+ * SeaweedFS storage client binding for YCSB.
+ */
+package site.ycsb.db;
+
+import seaweedfs.client.*;
```
Comment (busbey) [inline seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java:21]: avoid wildstar imports please

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+   * Called once per SeaweedFS instance; If the client is not null it is re-used.
+   */
+  @Override
+  public void init() throws DBException {
+    filerHost = getProperties().getProperty("seaweed.filerHost", "localhost");
+    filerPort = Integer.parseInt(getProperties().getProperty("seaweed.filerPort", "8888"));
+    folder = getProperties().getProperty("seaweed.folder", "/ycsb");
+    filerGrpcClient = new FilerGrpcClient(filerHost, filerPort+10000);
+    filerClient = new FilerClient(filerGrpcClient);
+    filerClient.mkdirs(this.folder, 0755);
```
Comment (busbey) [inline seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java:92]: this is going to happen once per YCSB client thread. just want to make sure that's okay?

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+                              .setMtime(now)
+                              .setFileMode(0755)
+              );
+
+      SeaweedWrite.writeData(entry, "000", this.filerGrpcClient, 0, jsonData, 0, jsonData.length);
+
+      SeaweedWrite.writeMeta(this.filerGrpcClient, this.folder + "/" + tableName, entry);
+
+    } catch (Exception e) {
+      System.err.println("Not possible to write the object " + key);
```
Comment (busbey) [inline seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java:194]: use slf4j instead of System.err

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+   *
+   * @param tableName The name of the table
+   * @param key  The file key of the file to write.
+   * @param values A HashMap of field/value pairs to update in the record
+   * @return OK on success, ERORR otherwise.
+   */
+  @Override
+  public Status update(String tableName, String key,
+             Map<String, ByteIterator> values) {
+    return writeToStorage(tableName, key, values);
```
Comment (busbey) [inline seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java:146]: are the backend `SeaweedWrite.writeData` and/or `SeaweedRead.read` calls going to take care of coalescing with existing values?

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ * <p>
+ * SeaweedFS storage client binding for YCSB.
+ */
+package site.ycsb.db;
```
Comment (busbey) [inline seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java:19]: please put this in a package `site.ycsb.db.seaweed`

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java (hunk 6 above)
Comment (chrislusf) [reply seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java:92]: this is not ideal but ok. Could not find a better place to setup test initialization..

File: pom.xml (hunk 1 above)
Comment (chrislusf) [reply pom.xml]: My mistake. Removed.

btw: the S3 implementation seems wrong though.

File: seaweedfs/README.md (hunk 2 above)
Comment (chrislusf) [reply seaweedfs/README.md]: changed to 2020

File: seaweedfs/README.md (hunk 3 above)
Comment (chrislusf) [reply seaweedfs/README.md:18]: Added one liner of intro and link to the github repo.

File: seaweedfs/pom.xml (hunk 4 above)
Comment (chrislusf) [reply seaweedfs/pom.xml]: changing to 2020

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java (hunk 5 above)
Comment (chrislusf) [reply seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java]: fixed now.

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java (hunk 7 above)
Comment (chrislusf) [reply seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java]: done

File: seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java (hunk 8 above)
Comment (chrislusf) [reply seaweedfs/src/main/java/site/ycsb/db/SeaweedClient.java]: No. I added logic to merge the new value with existing value.

**Comment by chrislusf** [comment]
Somehow the Travis CI build failed with the following reason. Is there anything changed for YCSB Travis CI?

```
$ ./mvnw install -q -DskipTests=true
[INFO] Skipping plugin execution
[INFO] Skipping plugin execution
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:2.5.5:single (default) on project ycsb: Failed to create assembly: Artifact: site.ycsb:seaweedfs-binding:jar:0.18.0-SNAPSHOT (included by module) does not have an artifact with a file. Please ensure the package phase is run before the assembly is generated. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :ycsb
The command "./mvnw install -q -DskipTests=true" failed and exited with 1 during .
```

**Comment by chrislusf** [comment]
@busbey All requested changes have been completed already.

**Review by busbey — APPROVED** [review]
